lib/oli_web/controllers/cognito_controller.ex

```diff
@@ -0,0 +1,143 @@
+defmodule OliWeb.CognitoController do
+  use OliWeb, :controller
+
+  import Oli.{HTTP, Utils}
+
+  require Logger
+
+  alias Oli.{Repo, Sections, Accounts, Groups, Institutions}
+  alias Oli.Institutions.SsoJwk
+  alias Oli.Delivery.Sections
+  alias Oli.Delivery.Sections.Section
+
+  @jwks_relative_path "/.well-known/jwks.json"
+
+  def launch(
+        conn,
+        %{
+          "product_id" => product_id,
+          "cognito_id_token" => jwt,
+          "error_url" => _error_url,
+          "community_id" => community_id
+        } = params
+      ) do
+    with {:ok, jwk} <- get_jwk(jwt),
+         {true, %{fields: jwt_fields}, _} <- JOSE.JWT.verify_strict(jwk, ["RS256"], jwt),
+         %Section{} = section <- Sections.get_section_by(slug: product_id),
+         {:ok, user} <- setup_lms_user(jwt_fields, community_id) do
+      conn
+      |> use_pow_config(:user)
+      |> Pow.Plug.create(user)
+      |> redirect(to: redirect_path(conn, user, section.id))
+    else
+      nil ->
+        redirect_with_error(conn, params, "Invalid product")
+
+      {false, _, _} ->
+        redirect_with_error(conn, params, "Unable to verify credentials")
+
+      {:error, error} ->
+        redirect_with_error(conn, params, snake_case_to_friendly(error))
+
+      {:error, %Ecto.Changeset{}} ->
+        redirect_with_error(conn, params, "Invalid parameters")
+    end
+  end
+
+  def launch(conn, params) do
+    redirect_with_error(conn, params, "Missing parameters")
+  end
+
+  defp get_jwk(jwt) do
+    with {:ok, %{"kid" => kid}} <- Joken.peek_header(jwt),
+         {:ok, %{"iss" => issuer}} <- Joken.peek_claims(jwt),
+         {:ok, sso_jwk} <- find_or_fetch_jwt(kid, issuer) do
+      {:ok, JOSE.JWK.from_pem(sso_jwk.pem)}
+    else
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  def find_or_fetch_jwt(kid, issuer) do
```

Comment:
Nit: I realized this and some other functions below are not private. Is there any specific reason? If so I would move all the private ones to the bottom.